sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -758,777 +816,1070 @@ namespace Azure { namespace Core { namespace Http { namespace _detail {
-    // it's ok to initiate the request without a client certificate.
-    //
-    // Note: If/When TLS client certificate support is added to the pipeline, this line may need to
-    // be revisited.
-    if (!WinHttpSetOption(
-            m_requestHandle.get(),
-            WINHTTP_OPTION_CLIENT_CERT_CONTEXT,
-            WINHTTP_NO_CLIENT_CERT_CONTEXT,
-            0))
+    // If port is 0, i.e. INTERNET_DEFAULT_PORT, it uses port 80 for HTTP and port 443 for
+    // HTTPS.
+    uint16_t port = url.GetPort();
+
+    // Before doing any work, check to make sure that the context hasn't already been cancelled.
+    context.ThrowIfCancelled();
+
+    // Specify an HTTP server.
+    // This function always operates synchronously.
+    Azure::Core::_internal::UniqueHandle<HINTERNET> rv(WinHttpConnect(
+        m_sessionHandle.get(),
+        StringToWideString(url.GetHost()).c_str(),
+        port == 0 ? INTERNET_DEFAULT_PORT : port,
+        0));
+
+    if (!rv)
     {
-      GetErrorAndThrow("Error while setting client cert context to ignore.");
+      // Errors include:
+      // ERROR_WINHTTP_INCORRECT_HANDLE_TYPE
+      // ERROR_WINHTTP_INTERNAL_ERROR
+      // ERROR_WINHTTP_INVALID_URL
+      // ERROR_WINHTTP_OPERATION_CANCELLED
+      // ERROR_WINHTTP_UNRECOGNIZED_SCHEME
+      // ERROR_WINHTTP_SHUTDOWN
+      // ERROR_NOT_ENOUGH_MEMORY
+      GetErrorAndThrow("Error while getting a connection handle.");
     }
+    return rv;
   }
 
-  if (!options.ProxyInformation.empty())
+  void WinHttpRequest::EnableWebSocketsSupport()
   {
-    WINHTTP_PROXY_INFO proxyInfo{};
-    std::wstring proxyWide{StringToWideString(options.ProxyInformation)};
-    proxyInfo.dwAccessType = WINHTTP_ACCESS_TYPE_NAMED_PROXY;
-    proxyInfo.lpszProxy = const_cast<LPWSTR>(proxyWide.c_str());
-    proxyInfo.lpszProxyBypass = WINHTTP_NO_PROXY_BYPASS;
-    if (!WinHttpSetOption(
-            m_requestHandle.get(), WINHTTP_OPTION_PROXY, &proxyInfo, sizeof(proxyInfo)))
+#pragma warning(push)
+    // warning C6387: _Param_(3) could be '0'.
+#pragma warning(disable : 6387)
+    if (!WinHttpSetOption(m_requestHandle.get(), WINHTTP_OPTION_UPGRADE_TO_WEB_SOCKET, nullptr, 0))
+#pragma warning(pop)
     {
-      GetErrorAndThrow("Error while setting Proxy information.");
+      GetErrorAndThrow("Error while Enabling WebSocket upgrade.");
     }
   }
-  if (options.ProxyUserName.HasValue() || options.ProxyPassword.HasValue())
-  {
-    if (!WinHttpSetCredentials(
-            m_requestHandle.get(),
-            WINHTTP_AUTH_TARGET_PROXY,
-            WINHTTP_AUTH_SCHEME_BASIC,
-            StringToWideString(options.ProxyUserName.Value()).c_str(),
-            StringToWideString(options.ProxyPassword.Value()).c_str(),
-            0))
+
+  WinHttpRequest::WinHttpRequest(
+      Azure::Core::_internal::UniqueHandle<HINTERNET> const& connectionHandle,
+      Azure::Core::Url const& url,
+      Azure::Core::Http::HttpMethod const& method,
+      PCCERT_CONTEXT tlsClientCertificate,
```

Comment:
Why isn't the clientCertificate passed in via the WinHttpTransportOptions?

Reply:
Because at the time this is called, the PCCERT_CONTEXT might have been freed as soon as the WinHttpTransport is created.

I've added some comments about this, and zeroed the pointer out to make it clear that it cannot be relied upon.